design/pv-migration.md

```diff
@@ -0,0 +1,86 @@
+
+## Problem
+
+The problem with the LocalPV is, it has the affinity set on the PersistenceVolume object. This will let k8s scheduler to schedule the pods to that node as data is there only. ZFS-LocalPV driver uses nodename to set the affinity which creates the problem here as if we are replacing a node, the node name will change and k8s scheduler will not be able to schedule the pods to the new node even if we have moved the disks there.
+
+## Current Solution
+
+Current solution depends on `openebs.io/nodeid` node label. Admin has to set the same label to the replaced node which allows the k8s scheduler to automatically schedule the old pods to this new node. We can read more about the solution [here](https://github.com/openebs/zfs-localpv/blob/master/docs/faq.md#8-how-to-migrate-pvs-to-the-new-node-in-case-old-node-is-not-accessible).
+
+The problem with the above approach is we can not move the volumes to any existing node as we can set only one label on the node for a given key. So, for any existing node in a cluster, label `openebs.io/nodeid` will already be set so, we can not update it to a new value as the pods already running on it will not able to get scheduled here.
+
+
+## Proposal
+
+### Keys Per ZPOOL
+
+We are proposing to have a key dedicated to ZFS POOL. This key will be used by the ZFS-LocalPV driver to set the label on the nodes where it is present. In this way we can allow the ZFS POOLs to move from any node to any other node as the key is tied to the ZFS POOL as opposed to keeping it per node. We are proposing to have a `openebs.io/<poolname>=true` label on the node where the pool is present. Assuming admins do not have large number of pools on a node, there will be not much label set on a node.
+
+### Migrator
+
+ZFS POOL name should be same across all the nodes for ZFS-LocalPV. So, we have to rename the ZFS POOLs if we are moving it to any existing node. We need a Migrator workflow to update the POOL name in the ZFSVolume object. This will find all the volumes present in a ZFS POOL on that node and updates the ZFSVolume object with the correct PoolName.
+
+**Note:** We can not edit PV volumeAttributes with the new pool name as it is immutable field.
+
+The migrator will look for all the volumes start with the name "pvc-" and will look for corresponding ZFSVolume object and will update it. In case of admin has imported the existing volumes where it is possible that volume name might not start with the name "pvc-", the PoolName needs to be updated manually as the volume was created manually by the admin.
```

Comment:
To figure out `zvols` managed by CSI-driver can we set some customer ZFS property? So that it will reduce some manual steps.

Reply:
That manual step is still needed after that as volume is imported by admin manually.

Reply:
Yes, that is true admin needs to manually import the volume into CSI-Driver and it should be one-time activity. IMO for the sake of completeness, it is something like If the pool disk is moved to a different node then our CSI-Driver should take care of all the volumes which were previously managed by it irrespective of whether it is manually added by admin (or) created by K8s. 

> In case of admin has imported the existing volumes where it is possible that volume name might not start with the name "pvc-", the PoolName needs to be updated manually as the volume was created manually by the admin. 

To avoid the above manual step, Once the volume is handover to the CSI-Driver, we can set some custom `zfs property` (zfs set <openebs.io:is_managed=true>) indicating volume is managed by OpenEBS CSI-Driver and it will prevent the above additional manual steps when the disk is migrated to a different node.

Reply:
@mittachaitu I would not recommend to set anything on the zpool as we don't touch anything there and would like to keep it that way. Now coming to this point, we already have ZFSVolume CR, using that we can know that which volumes are managed by ZFSLocalPV driver.

Reply:
Ok, I am trying to propose to set a property on zfs/zvol volume dataset, not on the pool. But yeah, other way around is also possible i.e managing based on the ZFSVolume CR so that we can that manual step.